File: app/src/main/java/wangdaye/com/geometricweather/ui/widget/weatherView/materialWeatherView/IntervalComputer.java
package wangdaye.com.geometricweather.ui.widget.weatherView.materialWeatherView;

import android.content.Context;
import android.view.WindowManager;

public class IntervalComputer {

    private long mCurrentTime;
    private long mLastTime;

    private double mDefaultInterval;
    private double mInterval;

    public IntervalComputer(Context context) {
        reset(context);
    }

    public void reset(Context context) {
        mCurrentTime = -1;
        mLastTime = -1;

        double screenRefreshRate = 60;
        WindowManager windowManager = (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
        if (windowManager != null) {
            screenRefreshRate = windowManager.getDefaultDisplay().getRefreshRate();
        }
        if (screenRefreshRate < 60) {
            screenRefreshRate = 60;
        }

        mDefaultInterval = 1000.0 / screenRefreshRate;
        mInterval = mDefaultInterval;
    }

    public void invalidate() {
        mCurrentTime = System.currentTimeMillis();
        mInterval = mLastTime == -1 ? mDefaultInterval : (mCurrentTime - mLastTime);
        mLastTime = mCurrentTime;
    }

    public double getInterval() {
        return mInterval;
    }
}

File: app/src/main/java/org/breezyweather/theme/weatherView/materialWeatherView/IntervalComputer.kt
package org.breezyweather.theme.weatherView.materialWeatherView

class IntervalComputer {
    private var mCurrentTime: Long = 0
    private var mLastTime: Long = 0
    var interval = 0.0
        private set

    init {
        reset()
    }

    fun reset() {
        mCurrentTime = -1
        mLastTime = -1
        interval = 0.0
    }

    fun invalidate() {
        mCurrentTime = System.currentTimeMillis()
        interval = (if (mLastTime == -1L) 0 else mCurrentTime - mLastTime).toDouble()
        mLastTime = mCurrentTime
    }
}
